README.md
![mulTIColor Logo](https://repository-images.githubusercontent.com/474191033/6b4ec005-585c-4c27-a4f4-b6eea5f3b6ea)

mulTICcolor
===========

A converter for mulitcolor-mode on the [TIC-80](https://tic80.com/).

The [TIC-80](https://tic80.com/) normaly supports a max. [resolution](https://github.com/nesbox/TIC-80/wiki/display) of 240 pixels x 136 pixels with a [palette](https://github.com/nesbox/TIC-80/wiki/palette) of 16 RGB-colors.
But it is possible to change the palette every scanline with some [tweaks](https://github.com/nesbox/TIC-80/wiki/palette#more-than-16-colors), resulting in max. 2176 or even 4216 colors on screen.

**mulTICcolor** converts an image (240 x 136 pixels or smaller) to a version with max. 16 or 31 colors per line.
The result will be saved as an image and as a [Lua](https://www.lua.org)-script with all the palettes, pixels and a display-routine (when using the "o"-option). Images with 16 or less colors (indexed) will be converted too, but with a single palette.


Information
===========

To reduce the colors to 16/31 per line, there are 3 options by default to achieve this:
1. [PIL](https://pillow.readthedocs.io/en/stable/) (Python Image Library, built-in)
2. [IrfanView](https://www.irfanview.com/) (external executable)
3. [ImageMagick](https://imagemagick.org/index.php) (external executable)

You might have to adjust the paths for [IrfanView](https://www.irfanview.com/) and [ImageMagick](https://imagemagick.org/index.php) in the configfile (**mtc.cfg**).
You can also add other converters or paths for a different operating system to the config aswell. Make sure to include the three keywords **{IN}**, **{OUT}** and **{RANGE}** to your entries. 
They ensure, that the current line of the image which is written to a temporary file can be converted and imported again afterwards. There is an optional keyword **{TEMPDIR}**, which can be used to store additional files for a converter like [IrfanView](https://www.irfanview.com/), if needed.

Based on the source image, all available options should be tried and the best result can be used directly.
But it is recommended to manually edit/optimize the best result by the creator of the original image.

When using the -o option, a .[lua](https://www.lua.org)-script will be generated (too), which contains the palettes and pixel data.
The **pal**-variable will contain 136 palettes with 16 or 31 RGB-colors, each in hex (from 00 to ff).
The **gfx**-variable will contain 240 x 136 pixels (color-index in hex, from 0 to f with 16 colors / from 00 to 1e with 31 colors).
When using [RLE (Run-length encoding)](https://en.wikipedia.org/wiki/Run-length_encoding) the palettes and pixels are stored in the **rlp**- and **rlg**-variable.


Requirements
============

- Python (3.6, or greater) - https://www.python.org/
- Pillow (Python Imaging Library) - https://pypi.org/project/Pillow/


Installation
============
If missing, **Pillow** can be installed using **pip** (for Python3).

Linux: 

    $ pip install pillow
Windows:

    $ python -m pip install pillow


Quickstart
==========

**mulTIColor** is easy to use and its really simple to convert an image to multicolor-mode.
Just run `multicolor` using the provided example images:

    $ multicolor.py piggie_power_without_the_price.png

    Generates a preview of the converted image, will be shown by the OS's default image viewer.

Specify an output name to generate the lua-script and save the preview:

    $ multicolor.py kittens.png -o smallcats
	
	The preview is saved as "smallcats-mtc.png" and the lua-script as "smallcats.lua".

Use a different converter to reduce the colors of the image per line:

    $ multicolor.py dryad.png -c iview -o resultsmyvary
	
	Have a look at the provided config file "mtc.cfg" for the paths and binaries.

Switch the range to 31 colors per line to get a, hopefully, better result:

    $ multicolor.py kittens.png -c iview -r 31 -o somemorecolors

Encode the pixel data as rle (run-length encoded) to save some space:

    $ multicolor.py kittens.png -o smallerscript -m rle

Set a bordercolor to prevent streaking (will reduce colors per line to 15/30):

    $ multicolor.py gameboys.png -o blackborder -b 0 0 0


Commandline options
===================

    $ multicolor.py --help
	
    Usage: multicolor imagefile [OPTION...]
    Generate TIC-80 multicolor images
    
    mandatory arguments:
      imagefile          imagefile with graphicsdata (e.g.: .png, .gif, etc.)
    
    optional arguments:
      -c, --converter    converter: pil (default), see mtc.cfg for more
      -o, --output       outputfile for multicolor values (.lua)
      -f, --force        force overwrite of outputfile when it already exist
      -r, --range        range of colors per line (16 or 31)
      -b, --bordercolor  bordercolor (R G B) to prevent streaking
      -m, --mode         mode to encode values: raw (default) or rle
      -v, --version      show version info
      -h, --help         show this help
    
    The optional arguments are only needed if the default setting does not meet the
    required needs. A specific name for the outputfile can be set (-o / --output).
    The maximum range (-r / --range) of colors per line, can be set to 16 or 31.
    To prevent colorstreaking in the border, a specific color (-b / --bordercolor)
    can be set. WARNING: this will reduce the overall colors to 15 or 30 per line!
	Mode (-m / --mode) to encode the pixel data via rle (run-length encoding) or as
	raw, which is the default. To reduce the colors of the image per line, various
	converters (-c / --converter) can be used. These can be configured in "mtc.cfg".
	Additional defined are: "iview" for IrfanView, "magick" for ImageMagick.
    
    examples:
	  multicolor imagefile.png
      multicolor graphic.gif -o multicolor.lua
      multicolor pixels.png -c iview -o mydata.lua
      multicolor colorful.gif -r 16 -o only16.lua
      multicolor border.jpg -r 31 -b 255 127 64 -o frame.lua
      multicolor truecol.png -m rle -o compress.lua
      multicolor logo.gif -o overwriteme.lua -f


Files
=====

* **multicolor.py** (the converter itself)
* **images/piggie_power_without_the_price.png** (example image (23960 colors), [original](https://demozoo.org/graphics/205191/) by [Bossman](https://demozoo.org/sceners/32053/)/[Rift](https://www.pouet.net/groups.php?which=11428))
* **images/kittens.png** (example image (23704 colors), [original](https://demozoo.org/graphics/302070/) by [Evil](https://demozoo.org/sceners/5794/)/[Accession](https://www.pouet.net/groups.php?which=1004))
* **images/dryad.png** (example image (17056 colors), [original](https://demozoo.org/graphics/266505/) by [Lycan](https://demozoo.org/sceners/21309/)/[LNX](https://www.pouet.net/groups.php?which=11760))
* **images/gameboys.png** (example image (18538 colors), [original](https://commons.wikimedia.org/wiki/File:Chiptune-Setup-Game-Boys.jpg) by [Lucius Kwok](https://www.flickr.com/photos/luciuskwok/))
* **images/replicaanxiety.png** (example image (16 indexed colors), [original](https://demozoo.org/graphics/3719/) by [Decca](https://demozoo.org/sceners/634/)/[Lego](https://www.pouet.net/groups.php?which=556))
* **components/display16.lua** (the display-routine for 16 colors per line)
* **components/display31.lua** (the display-routine for 31 colors per line)
* **components/display.lua** (the display-routine for 16 or less indexed colors)
* **components/rle-decoder16.lua** (the decoder when using rle-mode for 16 colors)
* **components/rle-decoder31.lua** (the decoder when using rle-mode for 31 colors)
* **components/rle-decoder.lua** (the decoder when using rle-mode for 16 or less indexed colors)

Disclaimer
==========

There is no warranty for the scripts or it's functionality, use it at your own risk.

The icon/logo consists of the [Sweetie 16](https://lospec.com/palette-list/sweetie-16)-palette © by [GrafxKid](https://grafxkid.tumblr.com/) & a [Shredder clipart](https://www.clipartmax.com/middle/m2H7N4d3G6N4b1H7_shredder-machine-icon-clipart-paper-office-shredders-shredder-machine-icon/) © by [Clipartmax](https://clipartmax.com/).


Bug tracker
===========

If you have any suggestions, bug reports or annoyances please report them to the issue tracker at https://github.com/RiftTeam/multicolor/issues


Contributing
============

Development of `mulTIColor` happens at GitHub: https://github.com/RiftTeam/multicolor


Changelog
2022-03-27  Decca (Philipp) & Mantratronic (Roceal)
		* Version 0.9
                * Support for 16 colors per line
		* Initial release (beta)

2023-01-15  Decca (Philipp) & Mantratronic (Roceal)
		* Version 1.0
		* Added support for 31 colors per line

2023-08-03  Decca (Philipp)
		* Version 1.1
                * Released to the public
		* Palette-data now also rle-compressed

2023-08-10  Decca (Philipp)
		* Version 1.2
                * Set a bordercolor to prevent streaking

2023-10-02  Decca (Philipp)
		* Version 1.3
                * Convert images with 16 or less colors
                * Fixed including of components


components/display16.lua
-- title:  Display
-- author: Mantratronic/Rift
-- desc:   show a mulTIColor image
-- script: lua

-- CODEBLOCK

function setPalette(p)
	for i=0,15 do
		for j=0,2 do
			poke(0x03FC0 + (i*3) + j,p[(i*3) + j+1])
		end
	end
end

--- high colour image test
hci_numpals = 0
hci_lp={}
hci_width = 0
hci_height = 0
hci_data={}
hci_image=""
function hci_init()
	local tnb=tonumber
	local i =0
	-- load palettes
	hci_numpals = #pal/96 -- pal length / 2 characters * 3 RGB * 16 colours = 96
	for i=1,hci_numpals do
		currentIP = {} 
		for j=0,47 do
			x=(i-1)*96 + j*2 + 1
			c1 = tnb(pal:sub(x,x),16)
			c2 = tnb(pal:sub(x+1,x+1),16)
			c = c2+16*c1
			table.insert(currentIP,c)	
		end
		table.insert(hci_lp,currentIP)
	end

	-- load image
	local width = #gfx / hci_numpals
	local height = hci_numpals
	trace("width x height:"..width.."x"..height)
    hci_width = width
    hci_height = height

	if width <= 128 and height <= 128 then
	
		for i=0,width*height do
            local x = i%width
            local y = i//width
			c= tnb(gfx:sub(i+1,i+1),16)
            poke4(0x8000 + (((x//8+y//8*16)*32)*2+(x%8)+(y%8)*8),c)
		end
	elseif width <=128 and height > 128 then
    	-- if height > 128 (same as above, just fills spritesheet too)
        local addr = 0x8000
		
		for i=0,width*height do
            local x = i%width
            local y = i//width
			c= tnb(gfx:sub(i+1,i+1),16)
            poke4(addr + (((x//8+y//8*16)*32)*2+(x%8)+(y%8)*8),c)
		end
	elseif width >128 and height <= 128 then
	    -- if width > 128, stores rightmost pixels in spritesheet
        local addr = 0x8000
        local offset = 0
		
		for i=0,width*height do
            local x = i%width
            local y = i//width
			c= tnb(gfx:sub(i+1,i+1),16)

            if x >= 128 then
                offset = 0x4000
                x = x-128
            else
                offset = 0
            end
            poke4(addr + offset + (((x//8+y//8*16)*32)*2+(x%8)+(y%8)*8),c)
		end
	elseif width >128 and height > 128 then
		-- if both are bigger
		-- load 128 x 128 in the tilesheet
		
        local addr = 0x8000
        local offset = 0
		for i=0,width*height do
            local x = i%width
            local y = i//width
			c= tnb(gfx:sub(i+1,i+1),16)

            if y >= 128 then
                offset = 0x4000
                local leftover = 128 -(width-128)
                y = y-128 + 8 * (x // leftover)
                x = (x % leftover) + (width-128)
            elseif x >= 128 then
                offset = 0x4000
                x = x-128
            else
                offset = 0
            end
            poke4(addr + offset + (((x//8+y//8*16)*32)*2+(x%8)+(y%8)*8),c)
		end
	end
    sync(0,0,true)
end
function hci_update(ft)
	-- load palettes into currentLP
	currentLP={}
	for i=0,135 do
		table.insert(currentLP,hci_lp[i+ 1])
	end
end
function hci_draw(ft)
	-- switch to correct bank etc
	-- draw sprites
	cls(0)

    if hci_width <=128 and hci_height <=128 then
        spr(0,hci_width,0,-1,1,0,0,hci_width//8,hci_height//8)
        -- pixel by pixel
        for i=0,hci_width*hci_height-1 do
            c= tonumber(gfx:sub(i+1,i+1),16)
            pix(i%hci_width,i//hci_width,c)
        end
    elseif hci_width <=128 and hci_height > 128 then
        -- if height > 128 (same as above, just copies from spritesheet too)
        spr(0,hci_width,0,-1,1,0,0,hci_width//8,16)
        spr(256,hci_width,128,-1,1,0,0,hci_width//8,hci_height//8-16)
        -- pixel by pixel
        for i=0,hci_width*hci_height-1 do
            c= tonumber(gfx:sub(i+1,i+1),16)
            pix(i%hci_width,i//hci_width,c)
        end
    elseif hci_width >128 and hci_height <= 128 then
        -- if width > 128
	    -- reads rightmost pixels from spritesheet
        spr(0,0,0,-1,1,0,0,16,hci_height//8)
        spr(256,128,0,-1,1,0,0,hci_width//8-16,hci_height//8)
        --[[ pixel by pixel
        for i=0,hci_width*hci_height-1 do
            c= tonumber(gfx:sub(i+1,i+1),16)
            pix(i%hci_width,i//hci_width,c)
        end
        --]]
    elseif hci_width >128 and hci_height > 128 then
        -- if both > 128
        spr(0,0,0,-1,1,0,0,16,16)
	    -- reads rightmost pixels from spritesheet
        spr(256,128,0,-1,1,0,0,hci_width//8-16,16)
        -- reads leftovers
        local normal_width = (hci_width//8)
        local leftover_width = (32-hci_width//8)
        for level=0,normal_width/leftover_width do
            spr(256+normal_width-16 + level*16, level*leftover_width*8,128,-1,1,0,0,leftover_width,1)
        end
    end

	-- make sure active scan is on
	activeScan=1
end

--- high colour effect test
hce_palettes={}
function hce_init()
	for i=0,20 do
		hce_palettes[i] = mixPalette2(palettes[3],aqua,i/20)
		--table.insert(hce_palettes,mixPalette(gray,aqua,i/20))
	end
end
function hce_update(ft)
	currentLP={}
	hce_palettes={}
	for i=0,20 do
		hce_palettes[i] = mixPalette2(palettes[6],aqua,i/20)
		--table.insert(hce_palettes,mixPalette(gray,aqua,i/20))
	end
	for i=0,135 do
		--[[
		if (i)//1%3 == 1 then
			table.insert(currentLP,palettes[1])
		else
			table.insert(currentLP,palettes[2])
		end--]]
		table.insert(currentLP,hce_palettes[(i+ft)//4%#hce_palettes])
		table.insert(currentLP,hce_palettes[(i+ft)//4%#hce_palettes])
	end
end
function hce_draw(ft)
	-- switch to the right bank, should probably only happen on scene switch
	sync(35,1,false)
	spr(0,0,0,-1,1,0,0,16,16)
	spr(256,128,1,-1,1,0,0,16,16)
	activeScan=1
	poke(0x3ff8,4)
end

activeScan=1
currentP={}
currentLP={}
function SCN(l)
	if activeScan == 1 then
		--[[]
		for i=0,47 do
			poke(0x03FC0 + i,(i*l)%255)
		end --]]
		setPalette(currentLP[(l%#currentLP)+1])
	elseif (l==0) then
		setPalette(currentP)
	end
end

hci_init()
function TIC()
--trace("pal length:"..(#pal/(6*16)).." | gfx length:"..(#gfx/240))
hci_update(1)
hci_draw(1)
end

components/rle-decoder.lua
-- title:  RLE Decoder
-- author: Decca/Rift
-- desc:   decode rle to global var
-- script: lua

-- DESCRIPTION
--
-- store pixel values (colors) as rle-encoded string
--  format: <rle-encoded colorvalues>
--     rle: colorvalues in uppercase letters - A=0,B=1...P=15
--          repeat of values in digits - 0...9
--          easier decoding using different symbols for values vs. repeats
--          a single value has no repeat-digit: A
--          two same values have no repeat-digit, just double letters: AA
--          three or more same values are appended with repeat-digits: A2
--          three: the letter itself and two additional repeats: A2 > AAA
--          seven: the letter itself and six additional repeats: A6 > AAAAAAA
local rle = "MK4M3PMMKKM2NMMCFM2KMMEPM2PMMKKMMK4M3KM4AK2AM2"

-- CODEBLOCK

-- the rle-decoder
function togfx(str)
  local d = "" -- (d)ecoded data
  for m, c in str:gmatch("(%u+)([^%u]+)") do -- decode rle, (m)atch & (c)ounter
    d = d .. m .. (m:sub(-1):rep(c))
  end
  _G.gfx = "" -- global graphic variable
  for v = 1,#d,1 do
    gfx = gfx .. string.format("%x",string.byte(d:sub(v,v))-65) -- get (c)olor value in hex
  end
end

-- call the decoder
togfx(rle)


multicolor.py
#!/usr/bin/python3
#
# mulTIColor - convert images to TIC-80 multicolor mode
#
# Hint: python -m pip install pillow (install PIL on Windows)
#
# last updated by Decca / RiFT on 02.10.2023 14:55


# import modules
from PIL import Image
import subprocess
import argparse
import tempfile
import os.path
import sys


# replace argparse-error message with own, nicer help/usage
class ArgumentParser(argparse.ArgumentParser):
    def error(self, message):
        print("Usage: multicolor imagefile [OPTION...]\n"
              "Generate TIC-80 multicolor images\n"
              "\n"
              "mandatory arguments:\n"
              "  imagefile          imagefile with graphicsdata (e.g.: .png, .gif, etc.)\n"
              "\n"
              "optional arguments:\n"
              "  -c, --converter    converter: pil (default), see mtc.cfg for more\n"
              "  -o, --output       outputfile for multicolor values (.lua)\n"
              "  -f, --force        force overwrite of outputfile when it already exist\n"
              "  -r, --range        range of colors per line (16 or 31)\n"
              "  -b, --bordercolor  bordercolor (R G B) to prevent streaking\n"
              "  -m, --mode         mode to encode values: raw (default) or rle\n"
              "  -v, --version      show version info\n"
              "  -h, --help         show this help\n"
              "\n"
              "The optional arguments are only needed if the default setting does not meet the\n"
              "required needs. A specific name for the output file can be set (-o / --output).\n"
              "The maximum range (-r / --range) of colors per line, can be set to 16 or 31.\n"
              "To prevent colorstreaking in the border, a specific color (-b / --bordercolor)\n"
              "can be set. WARNING: this will reduce the overall colors to 15 or 30 per line!\n"
              "Mode (-m / --mode) to encode the pixel data via rle (run-length encoding) or as\n"
              "raw, which is the default. To reduce the colors of the image per line, various\n"
              "converters (-c / --converter) can be used. These can be configured in \"mtc.cfg\".\n"
              "Additional defined are: \"iview\" for IrfanView, \"magick\" for ImageMagick.\n"
              "\n"
              "examples:\n"
              "  multicolor imagefile.png \n"
              "  multicolor graphic.gif -o multicolor.lua\n"
              "  multicolor pixels.png -c iview -o mydata.lua\n"
              "  multicolor colorful.gif -r 16 -o only16.lua\n"
              "  multicolor border.jpg -r 31 -b 255 127 64 -o frame.lua\n"
              "  multicolor truecol.png -m rle -o compress.lua\n"
              "  multicolor logo.gif -o overwriteme.lua -f\n", file=sys.stderr)
        self.exit(1, '%s: ERROR: %s\n' % (self.prog, message))


# set commandline arguments
parser = ArgumentParser(prog='mulTIColor', add_help=False)
parser.add_argument('image',
                    metavar='imagefile',
                    type=str,
                    help='imagefile with pixels')
parser.add_argument('-c', '--converter',
                    metavar='converter',
                    const='pil',
                    default='pil',
                    type=str,
                    nargs='?',
                    action='store',
                    help='converter to reduce colors per line')
parser.add_argument('-o', '--output',
                    metavar='outputfile',
                    type=str,
                    action='store',
                    help='outputfile for multicolor data')
parser.add_argument('-f', '--force',
                    action='store_true',
                    help='force overwrite of outputfile')
parser.add_argument('-r', '--range',
                    metavar='range',
                    const=16,
                    default=16,
                    choices=(16, 31),
                    type=int,
                    nargs='?',
                    action='store',
                    help='range of colors per line (16 or 31)')
parser.add_argument('-b', '--bordercolor',
                    metavar='255',
                    type=int,
                    nargs=3,
                    action='store',
                    help='bordercolor (R G B) to prevent streaking')
parser.add_argument('-m', '--mode',
                    metavar='mode',
                    const='raw',
                    default='raw',
                    choices=('raw', 'rle'),
                    type=str,
                    nargs='?',
                    action='store',
                    help='encode data in other formats than raw')
parser.add_argument('-v', '--version',
                    action='version',
                    version='%(prog)s 1.3')
args = parser.parse_args()


# get commandline arguments
imageFile = args.image
outputConv = args.converter
outputFile = args.output
outputEnc = args.mode
outputBorder = args.bordercolor
outputRange = args.range
outputForce = args.force


# set or init vars
configFile = "mtc.cfg"
iniFile = ""
convCommand = None
convBinary = None
if outputRange > 16:
    digits = 2
else:
    digits = 1


# load image file
print("       Image: " + imageFile)
try:
    orgImg = Image.open(imageFile)
except Exception as error:
    print("ERROR: " + str(error), file=sys.stderr)
    exit(1)


# get image mode & format
orgMode = orgImg.mode
orgFormat = orgImg.format
orgFormat = '.' + orgFormat.lower()
print("      Format: " + orgFormat + " (" + orgMode + ")")


# get image dimensions
orgSizeX, orgSizeY = orgImg.size
print("  Resolution: " + str(orgSizeX) + " x " + str(orgSizeY))


# get image colors & amount
orgColors = orgImg.convert('RGB').getcolors(maxcolors=(orgSizeX * orgSizeY))
dominantColor = max(orgColors)
print("      Colors: " + str(len(orgColors)))


# check and set bordercolor
if len(orgColors) > 16:
    if isinstance(outputBorder, list):
        borderColor = tuple(outputBorder)
        print(" Bordercolor: " + str(borderColor).replace("(", "").replace(")", ""))
        borderHex = ""
        for color in borderColor:
            borderHex = borderHex + str('%0*x' % (2, color))
        # reduce colorrange per line, because of the bordercolor
        workRange = outputRange - 1
    else:
        workRange = outputRange
    print("       Range: " + str(workRange) + " colors")
else:
    # empty outputRange when image is not multicolor
    workRange = outputRange
    outputRange = ""
    print("       Range: indexed colors")


# check image dimensions
if orgSizeX > 240:
    print("ERROR: image is wider than 240 pixels")
    exit(1)
if orgSizeY > 136:
    print("ERROR: image is higher than 136 pixels")
    exit(1)


# check and read config for external converter
if outputConv != 'pil':
    if not os.path.isfile(configFile):
        print("ERROR: no configuration file found (" + configFile + ")")
        exit(1)
    else:
        with open(configFile) as file:
            for line in file:
                if line.startswith(outputConv + " = "):
                    name, command = line.split('=', 1)
                    convCommand = command.strip()
                else:
                    continue
        # check and parse converter command
        if not isinstance(convCommand, str):
            print("ERROR: converter \"" + outputConv + "\" is not defined")
            exit(1)
        else:
            # check and replace keywords
            if not all(x in convCommand for x in ["{IN}", "{OUT}"]):
                print("ERROR: {IN} and/or {OUT} are missing")
                exit(1)
            else:
                tmpDir = (tempfile.gettempdir())
                # print(tmpDir)  # DEBUG
                inFile = os.path.join(tmpDir, 'mtc-infile' + orgFormat)
                outFile = os.path.join(tmpDir, 'mtc-outfile' + orgFormat)
                convSubproc = convCommand.replace('{IN}', '"' + inFile + '"')
                convSubproc = convSubproc.replace('{OUT}', '"' + outFile + '"')
                # Range & TempDir are optional, no need for checks
                convSubproc = convSubproc.replace('{RANGE}', '"' + str(workRange) + '"')
                convSubproc = convSubproc.replace('{TEMPDIR}', '"' + tmpDir + '"')
            # check converter binary
            convBinary = convCommand.split('"')[1].split('"')[0]
            if not os.path.isfile(convBinary):
                print("ERROR: " + convBinary + " not found")
                exit(1)


# create ini-file for Irfanview if needed
if convBinary is not None:
    # print(os.path.basename(convBinary))  # DEBUG
    if "i_view" in os.path.basename(convBinary):
        if "32" in os.path.basename(convBinary):
            iniFile = os.path.join(tmpDir, 'i_view32.ini')
        else:
            iniFile = os.path.join(tmpDir, 'i_view64.ini')
        try:
            with open(iniFile, 'w') as file:
                file.write("[Batch]\n")
                file.write("AdvUseBPP=1\n")
                file.write("AdvBPP=" + str(workRange) + "\n")
                file.write("AdvUseFSDither=0\n")
                file.write("AdvDecrQuality=1\n")
        except Exception as error:
            print("ERROR: " + str(error), file=sys.stderr)
            exit(1)


# remove file-extension for saner filenaming
if outputFile.endswith(".lua"):
    outputFile = outputFile[:-len(".lua")]


# function when using an external converter
def ext_conv(convLine):
    # save line to reduce
    try:
        convLine.save(inFile)
    except Exception as error:
        print("ERROR: " + str(error), file=sys.stderr)
    # execute converter
    subprocess.run(convSubproc, shell=True)
    # load reduced line
    try:
        doneLine = Image.open(outFile)
    except Exception as error:
        print("ERROR: " + str(error), file=sys.stderr)
    return doneLine


# function to convert a multicolor-image to range of colors per line
def convert_range():
    print("   Converting with " + outputConv + "...")
    offsetY = 0
    toReduce = 0
    alreadyOk = 0
    while offsetY < orgSizeY:
        Line = orgImg.crop((0, offsetY, orgSizeX, offsetY + 1))
        Colors = Line.convert('RGB').getcolors(maxcolors=(orgSizeX))
        if len(Colors) > workRange:
            if outputConv == 'pil':
                newLine = Line.convert("P", palette=Image.ADAPTIVE, colors=workRange)
            else:
                extLine = ext_conv(Line)
                newLine = extLine.convert("P", palette=Image.ADAPTIVE, colors=workRange)
            newImg.paste(newLine, (0, offsetY))
            toReduce = toReduce + 1
        else:
            newImg.paste(Line, (0, offsetY))
            alreadyOk = alreadyOk + 1
        offsetY = offsetY + 1
    print("  already ok: " + str(alreadyOk))
    print("     reduced: " + str(toReduce))
    # delete tempfiles if necessary
    if outputConv != 'pil':
        if os.path.isfile(inFile):
            os.remove(inFile)
        if os.path.isfile(outFile):
            os.remove(outFile)
        if os.path.isfile(iniFile):
            os.remove(iniFile)


# function to get values from a converted multicolor-image
def get_multicolor():
    Pal = ""
    Pix = ""
    offsetY = 0
    offsetX = 0
    while offsetY < orgSizeY:
        # set bordercolor
        if isinstance(outputBorder, list):
            Pal = Pal + borderHex
            palOffset = 1
        else:
            palOffset = 0
        Line = newImg.crop((0, offsetY, orgSizeX, offsetY + 1))
        rawLine = Line.convert("P", palette=Image.ADAPTIVE, colors=workRange)
        # pixel values in hex (0 to f)
        while offsetX < orgSizeX:
            palIndex = rawLine.getpixel((offsetX, 0))
            hexVal = '%0*x' % (digits, (palIndex + palOffset))
            Pix = Pix + hexVal
            offsetX = offsetX + 1
        offsetY = offsetY + 1
        offsetX = 0
        # palette in hex (00 to ff)
        rawPalette = rawLine.getpalette()
        rgbEntries = workRange * 3
        rgbPalette = rawPalette[:rgbEntries]
        for entry in rgbPalette:
            hexVal = '%0*x' % (2, entry)
            Pal = Pal + hexVal
    return (Pix, Pal)


# function to get values from an indexed-image
def get_indexed():
    # set palette for output file
    stdPalette = "1a1c2c5d275db13e53ef7d57ffcd75a7f07038b76425717929366f3b5dc941a6f673eff7f4f4f494b0c2566c86333c57"  # SWEETIE-16 palette
    newPalette = newImg.getpalette()  # get original palette from image
    rgbEntries = (2 ** 4) * 3
    rgbPalette = newPalette[:rgbEntries]
    Pal = ""
    for entry in rgbPalette:
        hexVal = '%0*x' % (2, entry)
        Pal = Pal + hexVal
    Pal = Pal + stdPalette[len(Pal):]  # fill up with default palette
    # swap values (low & high nibble) for the unified decoder
    Pal = ''.join([Pal[val:val + 2][::-1] for val in range(0, len(Pal), 2)])  # (https://stackoverflow.com/a/4606057)
    Pix = ""
    # get values from converted image
    offsetY = 0
    offsetX = 0
    while offsetY < orgSizeY:
        while offsetX < orgSizeX:
            palIndex = newImg.getpixel((offsetX, offsetY))
            hexVal = '%0*x' % (1, palIndex)
            Pix = Pix + hexVal
            offsetX = offsetX + 1
            # print(palIndex, hexVal)  # DEBUG
        # exit(1)  # DEBUG
        offsetY = offsetY + 1
        offsetX = 0
    return (Pix, Pal)


# start the conversion
print("   Generating data...")
if len(orgColors) > 16:
    # create new image for pasting
    newImg = Image.new(mode="RGB", size=(orgSizeX, orgSizeY))
    # convert lines to color-range
    convert_range()
    # get multicolor values
    Pixels, Palette = get_multicolor()
else:
    # check image mode and convert if neccessary
    if orgMode != 'P':
        newImg = orgImg.convert("P", palette=Image.ADAPTIVE, colors=len(orgColors))
        print("     WARNING: image converted to indexed colors, results may vary")
    else:
        newImg = orgImg
    # set digits to 16 colors only
    digits = 1
    # get indexed values
    Pixels, Palette = get_indexed()


# compress 16 colors pixel data (rle: only append number if value repeats more than twice)
def encode_rle16(data):
    enc = ""
    prev = ""
    count = 1
    for symbol in data:
        value = chr(int(symbol, 16) + 65)
        if value != prev:
            if prev:
                enc = enc + prev
                if count == 2:
                    enc = enc + prev
                elif count > 2:
                    enc = enc + str(count - 1)
            count = 1
            prev = value
        else:
            count = count + 1
    enc = enc + prev
    if count == 2:
        enc = enc + prev
    elif count > 2:
        enc = enc + str(count - 1)
    if not enc[-1].isdigit():
        enc = enc + "0"
    return enc


# compress 31 colors pixel data (rle: only append number if value repeats more than twice)
def encode_rle31(data):
    enc = ""
    prev = ""
    count = 1
    for char1, char2 in zip(data[::digits], data[digits - 1::digits]):
        value = chr(int(char1 + char2, 16) + 65)
        if value != prev:
            if prev:
                enc = enc + prev
                if count == 2:
                    enc = enc + prev
                elif count > 2:
                    enc = enc + str(count - 1)
            count = 1
            prev = value
        else:
            count = count + 1
    enc = enc + prev
    if count == 2:
        enc = enc + prev
    elif count > 2:
        enc = enc + str(count - 1)
    if not enc[-1].isdigit():
        enc = enc + "0"
    enc = enc.replace('\\', '&')
    return enc


# compress the data if needed
if outputEnc == 'rle':
    if digits == 1:
        RlePalette = encode_rle16(Palette)  # RLE encode palette (0-f/1 digit)
        RlePixels = encode_rle16(Pixels)  # RLE encode pixels (0-f/1 digit)
    if digits == 2:
        RlePalette = encode_rle16(Palette)  # RLE encode palette (still 0-f/1 digit)
        RlePixels = encode_rle31(Pixels)  # RLE encode pixels (31colors/2 digits)
else:
    RawPalette = Palette
    RawPixels = Pixels


# show final image when no outputfile
if not isinstance(outputFile, str):
    print("  try to show image...")
    newImg.show()
    print("  done.")
    exit()


# check if output file already exist
def check_file(outputName):
    if not outputForce:
        if os.path.isfile(outputName):
            print("ERROR: file already exist")
            exit(1)
    else:
        return


# include display component for the outputfile
displayFile = os.path.join(os.path.dirname(__file__), "components", "display" + str(outputRange) + ".lua")
if os.path.isfile(displayFile):
    with open(displayFile, "r") as file:
        fileLines = [line.strip('\n') for line in file.readlines()]
        codeStart = next((index for index, tag in enumerate(fileLines) if tag == '-- CODEBLOCK'), -1)
        compDisplay = "\n".join(fileLines[codeStart + 1:]) + "\n"
else:
    print("ERROR: no display component found")
    compDisplay = "\n-- ERROR: no display component found!\n"


# include decoder component for the outputfile
if outputEnc == "rle":
    decoderFile = os.path.join(os.path.dirname(__file__), "components", "rle-decoder" + str(outputRange) + ".lua")
    if os.path.isfile(decoderFile):
        with open(decoderFile, "r") as file:
            fileLines = [line.strip('\n') for line in file.readlines()]
            codeStart = next((index for index, tag in enumerate(fileLines) if tag == '-- CODEBLOCK'), -1)
            compDecoder = "\n".join(fileLines[codeStart + 1:]) + "\n"
    else:
        print("ERROR: no decoder component found")
        compDecoder = "\n-- ERROR: no decoder component found!\n"
else:
    compDecoder = ""


# save generated data to outputfile (lua-script)
dataFile = outputFile + ".lua"
print(" try to save: " + str(dataFile))
check_file(dataFile)
try:
    with open(dataFile, 'w') as file:
        file.write("-- title:  " + str(imageFile) + "\n")  # write header
        file.write("-- author: mulTIColor\n")
        file.write("-- script: lua\n")
        file.write("\n")
        if outputEnc == 'rle':
            file.write('rlp = "' + RlePalette + '"\n')  # write RLE palette
            file.write("\n")
            file.write('rlg = "' + RlePixels + '"\n')  # write RLE pixels
        else:
            file.write('pal = "' + RawPalette + '"\n')  # write palette (RAW)
            file.write("\n")
            file.write('gfx = "' + RawPixels + '"\n')  # write pixels (RAW)
        if len(orgColors) <= 16:
            file.write("\n")
            file.write('res = ' + str(orgSizeX - 1) + '\n')  # write resolution of lines
        file.write(compDecoder)
        file.write(compDisplay)
except Exception as error:
    print("ERROR: " + str(error), file=sys.stderr)
    exit(1)


# save converted image to outputfile (with prefix)
outputFile = outputFile + "-mtc"
print(" try to save: " + str(outputFile))
check_file(outputFile)
try:
    newImg.save(outputFile)
except ValueError as error:
    if 'unknown file extension' in str(error):
        print(" try to save: " + str(outputFile) + str(orgFormat))
        check_file(outputFile + orgFormat)
        try:
            newImg.save(outputFile + orgFormat)
        except Exception as error:
            print("ERROR: " + str(error), file=sys.stderr)
            exit(1)
    else:
        print("ERROR: " + str(error), file=sys.stderr)
        exit(1)
except Exception as error:
    print("ERROR: " + str(error), file=sys.stderr)
    exit(1)


# end message
print(" done.")


# end of code
exit()
